core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java

```diff
@@ -45,6 +55,14 @@
     private final PlannerContext plannerContext;
     private final StatementAnalyzerFactory statementAnalyzerFactory;
 
+    private final NonEvictableCache<CacheIdentityKey, ExpressionAnalyzer> analyzerCache = buildNonEvictableCache(
+            CacheBuilder.newBuilder()
+                    // Try to evict entries as soon as possible to keep cache relatively small
```

Comment:
IMO `TypeAnalyzer` should be created per query instance and possibly extracted as an interface. It should be then created and used in `io.trino.sql.planner.LogicalPlanner#plan`. No `NonEvictableCache<CacheIdentityKey, ExpressionAnalyzer> analyzerCache` is needed then